### PR TITLE
988 - Gnosis safe

### PR DESCRIPTION
**Pull request by hiaux0**
## What was done
- add gnosis safe package
`npm i @gnosis.pm/safe-apps-web3modal @gnosis.pm/safe-apps-sdk`
- add `manifest.json`
  Needed as described here:
  - https://docs.gnosis-safe.io/build-with-safe/sdks/safe-apps/get-started#manifest
  - More info here
    - https://docs.gnosis-safe.io/build-with-safe/sdks/safe-apps/releasing-your-safe-app#get-your-safe-app-listed-in-the-safe
- ~~add CORS to `vercel.json`~~ --> not needed actually -> reverted commit
  - needed: https://docs.gnosis-safe.io/build-with-safe/sdks/safe-apps/get-started#cors
  - vercel docs: https://vercel.com/support/articles/how-to-enable-cors#enabling-cors-using-vercel.json

#### Links for research
https://ethereum.stackexchange.com/questions/122635/signing-messages-as-a-gnosis-safe-eip1271-support
https://ethereum.stackexchange.com/questions/122458/web3-eth-personal-sign-with-walletconnect-and-gnosis-safe
https://ethereum.stackexchange.com/questions/117935/using-personal-sign-with-walletconnect-and-gnosis-safe/120440#120440=
https://ethereum.stackexchange.com/questions/124444/gnosissafe-signing-and-verifying-a-signature
https://ethereum.stackexchange.com/questions/128319/eip1271-signing-message-as-signer
https://ethereum.stackexchange.com/questions/124444/gnosissafe-signing-and-verifying-a-signature
https://ethereum.stackexchange.com/questions/116909/executing-transaction

### Setup
1. `npm run start-dev` (note, this PR adds (temporarily!) webpack devServer as https. Needed to expose manifest.json)
2. Make sure to rebuild firebase functions locally
3. Go to eg. https://gnosis-safe.io/app/eth:0x7450Aa5BCd1dc38A6Ac89e4D3BE976BE6AeBE9bc/home
4. Add Safe app (Apps -> Custom App -> https://localhost:3340)
5. Accept discliamer
6. Authenticate to firebase (new part, where a tx on Safe Wallet is created)

### Integration steps
https://www.notion.so/primedao/WIP-Gnosis-Safe-App-Instructions-2ef58ab2d719472185055430ff6de227#0760d7ee415d408da9fb38adddf7547c

### Testing the app
- Make sure wallet is behaving the same
```
- [x] connecting
  - [x] manually (connect)
  - [x] pops ups metamask
- [x] or automatically (wallet cached)
- [x] disconnecting has to work
- [x] Change account within metamask
```

#### Before
Wallet works

#### After
Wallet works
Appears in Gnosis Safe App

### CODE REVIEW

**Comment by hiaux0** [comment]
## Findings
#### Adding Custom App to Gnosis safe
- following docs https://help.gnosis-safe.io/en/articles/4022030-add-a-custom-safe-app
- manifest.json exists (on prod, this does not exists, neither is the Custom App recognized in Gnosis. This tells me, this branch is somehow working)
  - https://prime-deals-dapp-git-feat-988-gnosis-safe-curvelabs.vercel.app/manifest.json
image

File: firebaseFunctions/src/hash-validation.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+/**
+ * v1.3.0
+ * 1. https://rinkeby.etherscan.io/address/0x40597Caffbc904396DCAFD23786A0e1626E6975c#code=
+ * 2. "Similar Match Source Code"
+ * 2.1 https://rinkeby.etherscan.io/address/0x4cb09344de5bccd45f045c5defa0e0452869ff0f#readProxyContract
+ * 3. then expand "VERSION"
+ */
+const NEW_PROXY_VERSION = "1.3.0";
+
+const ProviderEndpoints = {
```
Comment (dkent600) [inline firebaseFunctions/src/hash-validation.ts:20]: Bummer to duplicate this from ethereumservice

File: firebaseFunctions/src/hash-validation.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+ */
+// const OLD_PROXY_VERSION = "1.1.1"
+/**
+ * v1.3.0
+ * 1. https://rinkeby.etherscan.io/address/0x40597Caffbc904396DCAFD23786A0e1626E6975c#code=
+ * 2. "Similar Match Source Code"
+ * 2.1 https://rinkeby.etherscan.io/address/0x4cb09344de5bccd45f045c5defa0e0452869ff0f#readProxyContract
+ * 3. then expand "VERSION"
+ */
+const NEW_PROXY_VERSION = "1.3.0";
```
Comment (dkent600) [inline firebaseFunctions/src/hash-validation.ts:18]: Why do you call this "proxy"?  Is it really?  Or is this better called something like "NEW_SAFE_CONTRACT_VERSION"

File: firebaseFunctions/src/index.ts [hunk 3]
```diff
@@ -238,10 +241,35 @@ export const CI = {
...
-        const signerAddress = verifyMessage(
-          message,
-          signature,
-        );
+        let signerAddress;
+        /**
+         * Verify the signature.
+         * There are 2 cases:
+         * 1. "Normal" signature
+         * 2. The WalletConnect Safe App uses EIP-1271
```
Comment (dkent600) [inline firebaseFunctions/src/index.ts:249]: Why do you say "WalletConnect"?  This is certainly needed when running as a Safe App, but that has nothing to do with WalletConnect.  Is it also needed when using a Safe Wallet from WalletConnect?  What about other wallets?

File: index.ejs [hunk 4]
```diff
@@ -12,6 +12,7 @@
   <link rel="stylesheet" href="./base.css">
   <link href="https://cdnjs.cloudflare.com/ajax/libs/slim-select/1.27.1/slimselect.min.css" rel="stylesheet">
   </link>
+  <link rel="manifest" href="manifest.json"/>
```
Comment (dkent600) [inline index.ejs:15]: should this be in a separate PR?  No big deal, but if this is required by Safe, maybe we can document it here.

**Review by dkent600 — CHANGES_REQUESTED** [review]

**Pull request merged**